### PR TITLE
[StructuralMechanics] Adding MainClass

### PR DESCRIPTION
**Pull request by philbucher**
This PR adds the ClassScript for StructuralMechanics, see #1278 

We have discussed this many many many times at the chair already and it is needed for FSI, Optimization and CoSimulation. (@AndreasWinterstein and I have this in production for our FSI for over half a year already)
The Interface follows what was discussed in #1278 and also especially the Interface of the CoSimulationApplication.

I am **leaving the classical MainScript**, so for a regular user nothing changes (although maybe at some point we can do what is already done in DEM)

The script is used in the tests so it is ensured that is working (since it has to be updated if sth changes otherwise the tests break)

### CODE REVIEW

**Review by loumalouomega — DISMISSED** [review]
Ok, thanks

**Review by loumalouomega — DISMISSED** [review]

**Comment by maceligueta** [comment]
Just a comment: it is very useful to add the typical 
```
if __name__ == "__main__":
    ClassStructuralMechanics("jsonfilename").Run()
```
or something similar at the end of the file. So you can run the script in the same fashion as before. Then, maybe you want to delete the previous main file. Up to you.

**Comment by philbucher** [comment]
thx @maceligueta I forgot to add it

**Comment by armingeiser** [comment]
The name of the file is a mixture of ```lower_case_with_underscores``` and ```UpperCamelCase```
Otherwhise i like it!

File: applications/StructuralMechanicsApplication/python_scripts/class_StructuralMechanics.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            KratosMultiphysics.Logger.PrintInfo("::[KSM Simulation]:: ", "Analysis -END- ")
+
+
+if __name__ == "__main__":
+    from sys import argv
+
+    if len(argv) > 2:
+        raise Exception("Wrong number of input arguments")
+
+    if len(argv) == 2: # ProjectParameters is being passed from outside
```
Comment (adityaghantasala) [inline applications/StructuralMechanicsApplication/python_scripts/class_StructuralMechanics.py]: May be also printing how to run and what are expected as arguments here can be helpful and informative ??

**Comment by philbucher** [comment]
@adityaghantasala this is what the error message says now:
~~~
Exception: Too many input arguments!
Use this script in the following way:
- With default ProjectParameters (read from "ProjectParameters.json"):
    "python3 class_structural_mechanics.py"
- With custom ProjectParameters:
    "python3 class_structural_mechanics.py CustomProjectParameters.json"
~~~

**Comment by adityaghantasala** [comment]
@philbucher  ... looks good .. !

**Comment by philbucher** [comment]
I plan to merge tomorrow such that we can use it soon

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            main_model_part_name = self.ProjectParameters["problem_data"]["model_part_name"].GetString()
+            self.main_model_part = KratosMultiphysics.ModelPart(main_model_part_name)
+            self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE,
+                                                      self.ProjectParameters["problem_data"]["domain_size"].GetInt())
+
+        ## Solver construction
+        import python_solvers_wrapper_structural
+        self.solver = python_solvers_wrapper_structural.CreateSolver(self.main_model_part, self.ProjectParameters)
+
+        if not using_external_model_part:
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py]: i believe you might want to AddVariable and AddDofs even for an external modelpart. At least in a future...

**Review by RiccardoRossi — DISMISSED** [review]
this looks fine to me, and i am approving since tests work on my computer.

My only feedback is that i think we should not yet consider as "Cast in stone" the public API of the StructuralMechanicsAnalysis object, in the sense that we shall be open to modifications to make it very similar to the equivalent one for the FluidDynamics and eventually for the DEM.

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py (hunk 2 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_analysis.py]: I discussed this extensively with @armingeiser but we haven't though abt what you are proposing.
It sounds good, I will discuss it with him tmr, thx for the suggestion though

**Comment by philbucher** [comment]
@RiccardoRossi yes this is not cast in stone, just a first step (based on the suggesions of @maceligueta , the requirements for CoSimulation and the FSI script @AndreasWinterstein and I use for quite a while now).
This reflects what is needed at the moment, but ofc we can change it later to make it consistent

**Comment by pooyan-dadvand** [comment]
Good work and thanks for providing a first version of solver class. :+1: 

Also I would add that after some discussion with @jcotela some time ago I'm in favor of having some more concrete methods like `ApplyBoundaryCondtion`, `DoSubStepping`, etc. Considering that this is the layer for our users (as optimizer our application developers) and it is not the best idea to keep our internal interface here giving methods that are not familiar to the users.

**Comment by RiccardoRossi** [comment]
@philbucher do you want to merge it for the release? if so do it NOW...

of the codacy warnings i think this one is pertinent:

https://app.codacy.com/app/miguelmaso/Kratos/pullRequest?prid=1404619

**Comment by philbucher** [comment]
@RiccardoRossi give me one moment I want to push to Interface funcitons

**Comment by philbucher** [comment]
@RiccardoRossi now this is ready for being merged

**Comment by philbucher** [comment]
Actually it is not needed for the release yet, only used in the tests...

**Comment by philbucher** [comment]
@pooyan-dadvand we are open to change the function names, what you propose makes also more sense to me.
Important to note at this point is **that this script reflects the requirements for Optimization, CoSimulation and FSI**

**Comment by jcotela** [comment]
@pooyan-dadvand if you want this for the release, I'm afraid you'll also need #1627

**Review by dbaumgaertner — COMMENTED** [review]
I guess a lot of functions in the "structural_mechanics_analysis.py" are private. Now that we agreed on a Python style guide I would recommend to actually apply it here. Meaning putting a leading "__" in front of the corresponding functions. This would make it clear what is supposed to be called from outside.

**Comment by RiccardoRossi** [comment]
+1 for @dbaumgaertner 's comment

**Comment by philbucher** [comment]
I discussed this with @dbaumgaertner and for our FSI cases those are public functions (In fact all of the functions that are currently there are public)

**Comment by RiccardoRossi** [comment]
@philbucher then i am in disagreement.
what is the public API right now? how is that general?

**Comment by RiccardoRossi** [comment]
we are removing this from the release branch, ok?  Idea is to merge at the beginning of the next cycle (tomorrow)

**Comment by philbucher** [comment]
@RiccardoRossi you are right, those are not the Public interface. My bad, fixed it

Also merging tmr is mine

**Review by dbaumgaertner — APPROVED** [review]

**Comment by dbaumgaertner** [comment]
We converged obviously, so I approved the PR. Before merging it, though, you might want to check the complaints of codacy again, whether these are actually meaningful.

**Comment by RiccardoRossi** [comment]
@philbucher ok thx!

in any case, could u summarize what is the expected usage and API? ideally in the wiki so that it remains easily accessible

**Comment by philbucher** [comment]
@dbaumgaertner : I checked codacy but the errors are not applicable (Application imports are needed and the typechecking I am doing because it is done like this everywhere in Kratos)

@RiccardoRossi briefly on the public Interface, I will also add it to the Wiki:
I followed the suggestions of @maceligueta and @adityaghantasala (Using the interface of the CoSimulationApp)
Public Interface:
- **Run** executes the entire problem
- **RunMainTemporalLoop** Executes the entire time loop
- **Initialize** Initializes the analysis
- **Finalize** Finalizes the analysis
- **InitializeTimeStep** Initializes the Timestep
- **SolveTimeStep** Solves the time step
- **FinalizeTimeStep**

I am merging now such that we can start using it :+1:

**Comment by dbaumgaertner** [comment]
Speaking for @KratosMultiphysics/optimization : Nice job!

**Pull request merged**